torch/testing/_internal/distributed/distributed_test.py

```diff
@@ -5123,52 +5123,67 @@ def test_output_unused_in_loss(self):
             for _ in range(10):
                 dist.all_reduce(torch.cat(tensors))
             # Run monitored barrier
-            timeout = 5
-            dist.monitored_barrier(None, timeout)
+            timeout = timedelta(seconds=2)
+            dist.monitored_barrier(timeout=timeout)
             # All ranks besides 1 call into barrier, rank 0 should report failure
             # while others report gloo error.
             failed_rank = 1
             if self.rank == failed_rank:
                 return
             if self.rank == 0:
                 with self.assertRaisesRegex(RuntimeError, f"Rank {failed_rank}"):
-                    dist.monitored_barrier(None, timeout)
+                    dist.monitored_barrier(timeout=timeout)
             else:
                 # It is permissible for other ranks that did not fail to
                 # successfully exit or crash in the monitored barrier, the main
                 # purpose of monitored barrier is to report the rank that hung
                 # on rank 0.
                 try:
-                    dist.monitored_barrier(None, timeout)
-                except RuntimeError as e:
+                    dist.monitored_barrier(timeout=timeout)
+                except RuntimeError:
                     pass
 
-        @require_backend({"gloo"})
-        @require_backends_available({"gloo"})
-        def test_monitored_barrier_gloo_subgroup(self):
-            failed_rank = 1
-            timeout = 5
-            subgroup = dist.new_group(ranks=[0, 1])
-            if self.rank == failed_rank:
+        @require_backend({"gloo", "nccl"})
+        @require_backends_available({"gloo", "nccl"})
+        def test_monitored_barrier_allreduce_hang(self):
+            # tests expected behavior when nonzero rank hangs.
+            if "NCCL_ASYNC_ERROR_HANDLING" in os.environ:
+                del os.environ["NCCL_ASYNC_ERROR_HANDLING"]
+            os.environ["NCCL_BLOCKING_WAIT"] = "1"
+            nccl_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                timeout=timedelta(seconds=10),
+                backend=dist.Backend.NCCL,
+            )
+            gloo_pg = dist.new_group(
+                ranks=list(i for i in range(int(self.world_size))),
+                backend=dist.Backend.GLOO,
+            )
+            tensors = [
+                torch.ones(10, device=self.rank) * self.rank
+            ]
+            # Let all ranks call allreduce first to set up communicators etc.
+            # Directly simulating error here will run into store issue described
+            # in https://github.com/pytorch/pytorch/issues/54524.
+            nccl_pg.allreduce(tensors).wait()
+            # All ranks besides 0 call into allreduce.
+            if self.rank != 0:
+                with self.assertRaisesRegex(RuntimeError, "Caught collective operation timeout"):
+                    nccl_pg.allreduce(tensors).wait()
                 return
```

Comment:
Why do we need to do this? Isn't it sufficient to just test that the monitored barrier throws for a subgroup?

Reply:
This is weird, I think this commit got messed up due to the hg restack issues internally, and a bunch of the tests went missing. Will send out the correct version

Reply:
Yes, I'm not sure why this is rendering in this way but I didn't intend to remove this test. Think something got messed up while rebasing